File: CodingTest/성유진/22주차/비밀 코드 해독.java
import java.util.*;

class Solution {
  TreeSet<Integer> code = new TreeSet<>();
  int N, M;
  int answer = 0;
  int[][] qq;
  int[] ansCnt;
      
  public boolean check() {
    for (int i = 0; i < M; i++) {
      int cnt = 0;
      for (int j = 0; j < 5; j++) {
        if (code.contains(qq[i][j]))
          cnt++;
      }
      if (cnt != ansCnt[i])
        return false;
    }
    return true;
  }
      
  public void comb() {
    if (code.size() == 5) {
      if (check())
        answer++;
      return;
    }
    
    for (int i = 1; i <= N; i++) {
      if (code.isEmpty() || (!code.contains(i) && i > code.last())) {
        code.add(i);
        comb();
        code.remove(i);
      }
    }
  }
      
  public int solution(int n, int[][] q, int[] ans) {
    N = n;
    M = q.length;
    qq = q;
    ansCnt = ans;
    comb();
    return answer;
  }
}

File: CodingTest/성호/21주차/비밀코드해독.js
function solution(n, q, ans) {
  let answer = 0;
  let queue = [];

  function isPossible() {
    for (let i = 0; i < q.length; i++) {
      let cnt = 0;
      for (let j = 0; j < 5; j++) {
        for (let k = 0; k < 5; k++) {
          if (q[i][j] === queue[k]) {
            cnt++;
          }
        }
      }
      if (ans[i] !== cnt) {
        return false;
      }
    }
    return true;
  }

  function dfs(depth, idx) {
    if (depth === 5) {
      if (isPossible()) {
        answer++;
      }
      return;
    }
    for (let i = idx; i <= n; i++) {
      queue.push(i);
      dfs(depth + 1, i + 1);
      queue.pop();
    }
  }

  dfs(0, 1);

  return answer;
}
